### PR TITLE
service: hid: Access shared memory directly

### PR DESCRIPTION
**Pull request by german77**
Currently we have a duplicate of the shared memory. Each single change we do needs to be copied to ensure both locations have the same data. This made the service prone to data mismatches or delays.

On some edge cases a game could be reading wrong data. This PR fixes the issue by casting a pointer to the shared memory and eliminating all data duplication. This fixes the issues on my unit test and has the potential to fix many games. Specially the ones who had wrong state like #8162 or pokemon let's go not detecting controllers properly

### CODE REVIEW

**Comment by bunnei** [comment]
Untagging due to conflicts

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 1]
```diff
@@ -62,11 +62,14 @@ bool Controller_NPad::IsDeviceHandleValid(const Core::HID::SixAxisSensorHandle&
     return npad_id && npad_type && device_index;
 }
 
-Controller_NPad::Controller_NPad(Core::HID::HIDCore& hid_core_,
+Controller_NPad::Controller_NPad(Core::HID::HIDCore& hid_core_, u8* raw_shared_memory_,
                                  KernelHelpers::ServiceContext& service_context_)
     : ControllerBase{hid_core_}, service_context{service_context_} {
     for (std::size_t i = 0; i < controller_data.size(); ++i) {
+        assert(NPAD_OFFSET + ((i + 1) * sizeof(NpadInternalState)) < shared_memory_size);
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/npad.cpp:69]: static_assert

Comment (german77) [reply src/core/hle/service/hid/controllers/npad.cpp:69]: Tried static assert but failed. Compiler said "i" can't be evaluated as a constant

Comment (liamwhite) [reply src/core/hle/service/hid/controllers/npad.cpp:69]: You don't need to assert within the loop body; you can extract this one out of the loop to something like this:
```cpp
assert(NPAD_OFFSET + sizeof(NpadInternalState)*controller_data.size() <= shared_memory_size);
```

Comment (Morph1984) [reply src/core/hle/service/hid/controllers/npad.cpp:69]: And this can be made into `static_assert` as all of them are `constexpr`

File: src/core/hle/service/hid/controllers/npad.h [hunk 2]
```diff
@@ -188,6 +187,8 @@ class Controller_NPad final : public ControllerBase {
     static bool IsDeviceHandleValid(const Core::HID::VibrationDeviceHandle& device_handle);
 
 private:
+    static const std::size_t NPAD_COUNT = 10;
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/npad.h:190]:
```suggestion
    static constexpr std::size_t NPAD_COUNT = 10;
```

**Review by bunnei — APPROVED** [review]
